### PR TITLE
Cleanup dynamic params

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Re-organized some DynamicParameters to be similar to the proposed structure from he design doc](https://docs.google.com/document/d/1-QD8_7TZiq6CNRSMeSmyg0ChrSVfvxwt_e6eLqiqRDA/edit?usp=sharing) and removed a few that were unneeded

I also deleted the SimulatorBackend since it's not needed, and I didn't want to update its parameters :sweat_smile: 

### Testing Done

existing tests pass

### Resolved Issues

None, just trying to make our lives a bit easier.

### Length Justification

I mostly went over length due to deleting the SimulatorBackend. The rest is just a lot or renaming.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/backend/grsim_backend.cpp [hunk 1]
```diff
@@ -6,15 +6,16 @@
 
 const std::string GrSimBackend::name = "grsim";
 
-GrSimBackend::GrSimBackend()
-    : network_input(SSL_VISION_DEFAULT_MULTICAST_ADDRESS, SSL_VISION_MULTICAST_PORT,
-                    SSL_GAMECONTROLLER_MULTICAST_ADDRESS,
-                    SSL_GAMECONTROLLER_MULTICAST_PORT,
+GrSimBackend::GrSimBackend(std::shared_ptr<const NetworkConfig> network_config)
+    : network_config(network_config),
+      network_input(network_config->VisionIP()->value(),
```
Comment (akhilveeraghanta) [inline src/software/backend/grsim_backend.cpp]: I think we should rename this to `VisionIPv4Address` to be explicit.

File: src/software/backend/grsim_backend.cpp [hunk 2]
```diff
@@ -6,15 +6,16 @@
...
+                    network_config->VisionPort()->value(),
+                    network_config->GamecontrollerIP()->value(),
```
Comment (akhilveeraghanta) [inline src/software/backend/grsim_backend.cpp]: `GamecontrollerIP` -> `GamecontrollerIPv4Address`

**Review by akhilveeraghanta — COMMENTED** [review]
nice 💯

File: src/software/parameter/config/ai_control.yaml [hunk 3]
```diff
@@ -34,74 +30,14 @@ AIControlConfig:
...
-      default: 10.0
-      description: >-
-        When ignoreInvalidCameraData is true, any robot or ball detection with an x-coordinate greater than this
-        value is ignored.
+  OverrideRefboxPlay:
+    type: "bool"
+    default: false
+    description: >-
+        Overrides the current refbox play with the play
+        specified by currentPlay parameter
```
Comment (kelvinkoon) [inline src/software/parameter/config/ai_control.yaml]: What does `currentPlay` mean? Is it in this `ai_control.yaml` file? Or is it referring to `CurrentRefboxPlay` below?

File: src/software/parameter/config/network.yaml [hunk 4]
```diff
@@ -12,3 +12,26 @@ NetworkConfig:
...
+      default: "224.5.23.2"
+      type: "std::string"
+      description: The IP address to listen on for vision data
+    VisionPort:
+      min: 1
+      max: 65535
+      default: 10020
+      type: "int"
+      description: The port to listen on for vision data
+    GamecontrollerIP:
```
Comment (kelvinkoon) [inline src/software/parameter/config/network.yaml]: Going along with what @akhilveeraghanta 's comment said, consider changing to `GamecontrollerIPv4Address` here too.

File: src/software/parameter/config/network.yaml [hunk 5]
```diff
@@ -12,3 +12,26 @@ NetworkConfig:
...
+      default: "224.5.23.1"
+      type: "std::string"
+      description: The IP address to listen on for gamecontroller data
```
Comment (kelvinkoon) [inline src/software/parameter/config/network.yaml]: And likewise, if changing to `GamecontrollerIPv4Address`, change the description to specify "IPv4 address"

**Review by kelvinkoon — COMMENTED** [review]
Looks good overall! Left a few comments.

File: src/software/backend/grsim_backend.cpp (hunk 1 above)
Comment (MathewMacDougall) [reply src/software/backend/grsim_backend.cpp]: done

File: src/software/backend/grsim_backend.cpp (hunk 2 above)
Comment (MathewMacDougall) [reply src/software/backend/grsim_backend.cpp]: done

File: src/software/parameter/config/network.yaml (hunk 4 above)
Comment (MathewMacDougall) [reply src/software/parameter/config/network.yaml]: done

File: src/software/parameter/config/network.yaml (hunk 5 above)
Comment (MathewMacDougall) [reply src/software/parameter/config/network.yaml]: updated descriptions for both vision and gamecontroller

File: src/software/parameter/config/ai_control.yaml (hunk 3 above)
Comment (MathewMacDougall) [reply src/software/parameter/config/ai_control.yaml]: It's the `currentAIPlay`. Good catch, clarified

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/software/backend/grsim_backend.h [hunk 6]
```diff
@@ -3,13 +3,15 @@
...
+#include "software/parameter/dynamic_parameters.h"
 #include "software/world/world.h"
 
 class GrSimBackend : public Backend
 {
    public:
     static const std::string name;
-    GrSimBackend();
+    GrSimBackend(std::shared_ptr<const NetworkConfig> network_config =
+                     DynamicParameters->getNetworkConfig());
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/backend/grsim_backend.h:14]: This won't work with the generic factory (although I agree with you). I've made a note to discuss parameter injection with factories separately.

File: src/software/backend/wifi_backend.cpp [hunk 7]
```diff
@@ -8,15 +8,20 @@
...
-                    DynamicParameters->getAIControlConfig()->getRefboxConfig(),
-                    DynamicParameters->getCameraConfig()),
+                    DynamicParameters->getSensorFusionConfig()),
       ssl_proto_client(boost::bind(&Backend::receiveSSLWrapperPacket, this, _1),
-                       boost::bind(&Backend::receiveSSLReferee, this, _1))
+                       boost::bind(&Backend::receiveSSLReferee, this, _1),
+                       network_config->VisionIPv4Address()->value(),
+                       network_config->VisionPort()->value(),
+                       network_config->GamecontrollerIPv4Address()->value(),
+                       network_config->GamecontrollerPort()->value())
```
Comment (jonathanlew) [inline src/software/backend/wifi_backend.cpp]: Why pass in individual values? Why not pass in a whole config?

File: src/software/gui/visualizer/widgets/ai_control.cpp [hunk 8]
```diff
@@ -100,23 +101,22 @@ void setupDefendingSideComboBox(Ui::AutoGeneratedMainWidget *widget,
...
         else if (text == "East")
         {
-            refbox_config->mutableOverrideRefboxDefendingSide()->setValue(true);
+            sensor_fusion_config->mutableOverrideRefboxDefendingSide()->setValue(true);
             // TODO: Confirm how East and West map to positive and negative sides
-            refbox_config->mutableDefendingPositiveSide()->setValue(false);
+            sensor_fusion_config->mutableDefendingPositiveSide()->setValue(false);
         }
         else if (text == "West")
         {
```
Comment (jonathanlew) [inline src/software/gui/visualizer/widgets/ai_control.cpp:116]: can we fix this in this PR? Shouldn't be too hard...

File: src/software/sensor_fusion/sensor_fusion.cpp [hunk 9]
```diff
@@ -87,10 +87,7 @@ void SensorFusion::updateWorld(const Referee &packet)
 {
     // TODO remove DynamicParameters as part of
     // https://github.com/UBC-Thunderbots/Software/issues/960
-    if (DynamicParameters->getAIControlConfig()
-            ->getRefboxConfig()
-            ->FriendlyColorYellow()
-            ->value())
+    if (DynamicParameters->getSensorFusionConfig()->FriendlyColorYellow()->value())
```
Comment (jonathanlew) [inline src/software/sensor_fusion/sensor_fusion.cpp]: Why not make SensorFusionConfig a field in SensorFusion?

File: src/software/sensor_fusion/sensor_fusion.cpp [hunk 10]
```diff
@@ -230,57 +217,21 @@ Team SensorFusion::createFriendlyTeam(const std::vector<RobotDetection> &robot_d
...
-                                  ->getRefboxConfig()
-                                  ->EnemyGoalieId()
-                                  ->value();
+    RobotId enemy_goalie_id =
+        DynamicParameters->getSensorFusionConfig()->EnemyGoalieId()->value();
     new_enemy_team.assignGoalie(enemy_goalie_id);
     return new_enemy_team;
 }
 
-bool SensorFusion::isCameraEnabled(const SSL_DetectionFrame &detection)
```
Comment (jonathanlew) [inline src/software/sensor_fusion/sensor_fusion.cpp:252]: Why is this being deleted?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
nice cleanup

File: src/software/backend/grsim_backend.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/backend/grsim_backend.h:14]: Sorry, made a note where? Is there an issue open that we could move this discussion to?

File: src/software/sensor_fusion/sensor_fusion.cpp (hunk 10 above)
Comment (MathewMacDougall) [reply src/software/sensor_fusion/sensor_fusion.cpp:252]: Now that Division B effectively has 1 camera, there's no need to be able to disable multiple. We already have functionality to limit the area of the field that we see by coordinates, which is more flexible than disabling entire cameras, so this was no longer really needed.

File: src/software/backend/grsim_backend.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/backend/grsim_backend.h:14]: Lets move discussion to #1452

File: src/software/gui/visualizer/widgets/ai_control.cpp (hunk 8 above)
Comment (MathewMacDougall) [reply src/software/gui/visualizer/widgets/ai_control.cpp:116]: Should be easy but I'm going to call that out of scope and fix it with my other Visualizer changes this week.

File: src/software/backend/wifi_backend.cpp (hunk 7 above)
Comment (MathewMacDougall) [reply src/software/backend/wifi_backend.cpp]: The Network.yaml config contains values that wouldn't be used, and I thought this was a low-enough level there wasn't a huge benefit in passing the config al the way down. I don't want to "overuse" them and put them too deep if not needed.

File: src/software/sensor_fusion/sensor_fusion.cpp (hunk 9 above)
Comment (MathewMacDougall) [reply src/software/sensor_fusion/sensor_fusion.cpp]: Originally was just moving params and not focusing on injecting them. Made it a field and replaced usages with the member variable.

File: src/software/sensor_fusion/sensor_fusion.cpp (hunk 10 above)
Comment (jonathanlew) [reply src/software/sensor_fusion/sensor_fusion.cpp:252]: I thought there were two cameras at Robocup 2019

File: src/software/backend/wifi_backend.cpp (hunk 7 above)
Comment (jonathanlew) [reply src/software/backend/wifi_backend.cpp]: this bit of code will need to be duplicated in almost every backend we make, so I'd like this to look tidier. For example, we could make an SSLProtoCommunicationConfig

**Review by jonathanlew — COMMENTED** [review]
replied to some comments

File: src/software/sensor_fusion/sensor_fusion.cpp (hunk 10 above)
Comment (MathewMacDougall) [reply src/software/sensor_fusion/sensor_fusion.cpp:252]: 2 cameras for Divison A, 1 camera for Division B.

File: src/software/backend/wifi_backend.cpp (hunk 7 above)
Comment (MathewMacDougall) [reply src/software/backend/wifi_backend.cpp]: done as suggested.

**Review by jonathanlew — DISMISSED** [review]
:100:

**Review by LiCody — DISMISSED** [review]
LGTM 👍

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada:

**Review by EvanMorcom — APPROVED** [review]
LGTM 
👯

**Review by kelvinkoon — APPROVED** [review]

**Review by jonathanlew — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]
:shipit:

**Pull request merged**